### ISSUE TITLE
Package management in POMDPs.jl family

### ISSUE THREAD

**Issue by juliohm** [JuliaPOMDP/POMDPs.jl#114]
What is the rationale behind having packages installed through `POMDPs.add()` instead of the usual `Pkg.clone(PKG_URL)`?

If the reason is only to avoid the release cycle checks and all the bureoucracy involved in tagging versions, I believe that `Pkg.clone()` is perfectly suited for the task. Maintaining a custom package manager can become a nightmare with new releases of Julia and if people decide to change the way METADATA works, etc.

Can you please share your thoughts on this?

**Comment by rejuvyesh** [comment]
AFAIU this is only a temporary stopgap measure and we are all waiting for `Pkg3` to land :smile: .

**Comment by zsunberg** [comment]
Now I remember! The reason is so that the user doesn't have to find the url. I think this is actually pretty useful because some solvers depend on others, and we only want to have urls hardcoded in one place, and we don't want to have to maintain urls in any documentation.

POMDPs.add is just a light wrapper around Pkg.clone(), so Pkg.update, etc still works.

We also might have decided it was a good idea to have a special mechanism because we might want to do something with versioning (#103), but this is complicated, so hopefully Pkg3 will come soon.

**Comment by juliohm** [comment]
My personal opinion is that the maintenance cost is much higher than the benefit in this case. A list of URLs in POMDP.jl README.md seems perfectly fine and more transparent maybe?

To give you feedback, this is the set of actions I took as a new user:

1. Found out POMDPs.jl cool stuff in pkg.julialang.org (or was introduced to it in CS238)
2. Learned about the multitude of solvers out there
3. Realized that they don't have tags
4. Cloned them
5. Missed the existence of `POMDPs.add()`
6. Read the docs more carefully later to get it working on my use case
7. Asked myself, why there is a custom manager? Do I have something broken in my installation because I didn't use `POMDPs.add()`?
8. Read the source code and realized it is just cloning

**Comment by mykelk** [comment]
Yep, just cloning going on. And thanks for your feedback. We spoke with Tony Kelman about this, and it sounds like Pkg3 is what we need, but it hasn't been rolled out yet.

**Comment by juliohm** [comment]
Thank you @mykelk, it is my pleasure. Feel free to close the issue if you plan to continue with the current approach before Pkg3 is released.

**Comment by rejuvyesh** [comment]
I guess we could still clarify some issues raised by @juliohm in the docs/FAQs?

**Comment by zsunberg** [comment]
Yes, we should definitely clarify that it does nothing more than clone in
the docstring and in the docs

**Comment by etotheipluspi** [comment]
Just to clarify, what @rejuvyesh is referring to are Julia environments in Pkg3 that allow local package management for julia projects. This seems ideal for our needs. 

While I agree with @juliohm that `Pkg.clone()` works perfectly fine form most cases, one major issue is version compatibility between the interface in `POMDPs.jl` which is tied to the official tagged release installed using `Pkg.add()`, and what is actually used in any given JuliaPOMDP package which would be installed using `Pkg.clone()`. Sometimes these do not match up. You can find more discussion on this in #103.

**Comment by juliohm** [comment]
Ideally all software should have releases, the fact that there are no
releases for the solvers is something that could also be rethink.

Releases enforce stability from the user side. Rolling release software is
hardly adopted by commercial products or control systems like the ones you
guys know much better than me. Aircraft collision, etc.

-Júlio

**Comment by mykelk** [comment]
Agreed. My understanding is that Pkg3 will address this. As soon as it is released, the plan is to use that instead of clone.

**Comment by juliohm** [comment]
I believe that all solvers in the POMDP family could be tagged. That way we can forget about incompatibility between POMDP interface and solver implementations. Specially now that releasing in Julia has become a straightforward operation on GitHub: https://github.com/attobot/attobot

No need to be a master in Gitlogy anymore, just use GitHub releases and the bot will automatically submit the PR to `METADATA.jl`.

**Comment by mykelk** [comment]
They can be tagged, but I'm not sure they can go easily into metadata. They have concerns about using acronyms in their namespace, unless the acronyms are well known.

**Comment by zsunberg** [comment]
I think they should be tagged, but not in METADATA (Maybe some of the most basic and stable ones like MCTS and DiscreteValueIteration could go in there. MCTS might be a well-known enough acronym). Simply tagging doesn't help that much though - the real challenge is matching up tags of solvers with tags of POMDPs. If we were to put them in METADATA, we would need to vastly improve testing and maybe have code review (which will, perhaps directly subtract from research time) because right now they are too experimental and there is sometimes a significant lag before a project makes it to METADATA

**Comment by juliohm** [comment]
@zsunberg why matching up the tags is not easy? Shouldn't the tags follow the same scheme that all other packages follow? Why not specify the version of POMDPs.jl in the REQUIRE file? Can you please elaborate?

**Comment by juliohm** [comment]
Sorry, I missed the comment about the acronyms, makes sense. They are strict about package naming, I see your point.

**Comment by zsunberg** [comment]
Meant to respond to this earlier. Does putting unregistered packages in REQUIRE have any effect?

**Comment by juliohm** [comment]
Hi Zach, which packages are unregistered? The solvers? Can't the solvers be registered like other packages? If the maintainer of a solver is not able to adjust his/her package to get it approved in METADATA, I would be skeptical to try the package myself. This is not necessarily the case, but getting the package approved in METADATA is an indication of code quality, robustness and stability. Most packages I tried that had no versions tagged were bugged, had no tests and crashed with a single deviation from basic usage.

I am not following the details of Pkg3 and that is why I cannot help much, sorry. Meanwhile, I would give some of the solvers a tag. The sooner a decision like this is made, the lesser is the effort for transitioning.

**Comment by mykelk** [comment]
@zsunberg you can't do that currently, see https://github.com/JuliaLang/julia/issues/7737

**Comment by zsunberg** [comment]
@juliohm , yeah, the solvers are unregistered.
Yeah @mykelk, that's what I thought.

Here's my final opinion: The ultimate solution to this is Pkg3. Our current system is adequate for now. There will be very few changes to the POMDPs.jl interface, so dependencies with that can be resolved manually. Moreover, it IS possible to make solvers dependent on a particular version of POMDPs since it is registered. Some level of quality is encouraged and communicated by having the build badges for all the solvers on the POMDPs.jl README.

The package namespace and METADATA are shared global resources. I would rather have those in charge of that focus on Pkg3 than use their resources to spot check the quality of our still-fairly-experimental packages.

If most of the other solver authors think we should register more packages in METADATA, I am happy to go along.

Also, it would be helpful to tag solvers (not in METADATA though) with semantic versioning, but I do not plan on taking the lead to make this a formal requirement or enforcing it until Pkg3 comes.

**Comment by zsunberg** [comment]
This exists: https://github.com/tbreloff/MetaPkg.jl I did not look at it closely, but it might be worth looking into. How does tbreloff have time to do all this stuff???

**Comment by juliohm** [comment]
tbreloff, timholy, ... these guys are unbelievable! :)

Closing this issue, thanks.

**Comment by etotheipluspi** [comment]
Does anyone have time to take a closer look at
https://github.com/tbreloff/MetaPkg.jl for this problem? 

It seems like it does what we want. @zsunberg?

**Comment by zsunberg** [comment]
For anyone looking at this - @etotheipluspi and I discussed this briefly on slack. I don't think MetaPkg is a perfect solution for us, and if we do #135, it will solve most of our problems. I think it's ok to have a simple package manager for the solvers only.

**Comment by juliohm** [comment]
What is the status of this issue? Should it be closed?
